### ISSUE TITLE
Prisma 5.0 Migration `findUnique` on `@unique` columns that are enums

### ISSUE THREAD

**Issue by cdaz5** [prisma/prisma#20227]
### Bug description

Given a schema:
```
enum FakeEnum {
  abc
  xyz
}

model Fake {
  id          String       @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  type        FakeEnum @unique
}
```
After migrating to 5.0 our `findUnique` call for the `Fake` model with `type` returns null. If i switch to `findFirst` it will return the correct row.

```
prisma.fake.findUnique({
  where: {
    type: FakeEnum.abc
  }
})

// this does not find the row but switching to findFirst does
```

### How to reproduce

have a model that has a `@unique` field that is an enum and run a `findUnique` query against that column. Apologies if you're looking for a full repro I can try to put that together as well when I have more time, just wanted to get this report in asap.

### Expected behavior

I would expect that given the schema and models above with a row in the database that has the `type` of `abc` a query like the below would return that row:

```
prisma.fake.findUnique({
  where: {
    type:  FakeEnum.abc
  }
})
```

### Prisma information

```prisma
enum FakeEnum {
  abc
  xyz
}

model Fake {
  id          String       @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  type        FakeEnum @unique
}
```

```ts
prisma.fake.findUnique({
  where: {
    type:  FakeEnum.abc
  }
})
```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: 19.7.0

### Prisma Version

```
5.0.0
```

**Comment by imwexpex** [comment]
Same for us, the following code in Prisma 5 returns NULL, while the previous version of Prisma returns the correct value
```
await prisma.geoConfig.findUnique({
          where: {
            country_state: {
              country: Country.US,
              state: State.AK,
            },
          },
        }),
```

Schema
```
model GeoConfig {
  uuid String @id @db.Uuid

  country Country
  state   State

  enabled Boolean @default(true)

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  @@unique([country, state])
}

enum Country {
  US
}

enum State {
  AK
  // other US States
}
```

**Comment by janpio** [comment]
So `findUnique` in Prisma 5 can not filter by enum values any more in your experience?

Do `Country.US` and `State.AK` still return the correct value as expected?
Can you turn on [query logging](https://pris.ly/d/logging) and [debugging](https://pris.ly/d/debugging) and post the full output you get please? Thanks.

**Comment by imwexpex** [comment]
@janpio 

`So findUnique in Prisma 5 can not filter by enum values any more in your experience?` - Yes

`Do Country.US and State.AK still return the correct value as expected?` - Yes, from generated prisma client:
```
export const Country: {
  US: 'US',
};
export const State: {
  AK: 'AK',
}
```

Query log:
```
prisma:query SELECT "public"."GeoConfig"."uuid", "public"."GeoConfig"."country", "public"."GeoConfig"."state", "public"."GeoConfig"."minAge", "public"."GeoConfig"."enabled", "public"."GeoConfig"."createdAt", "public"."GeoConfig"."updatedAt", "public"."GeoConfig"."updatedBy" FROM "public"."GeoConfig" WHERE (("public"."GeoConfig"."country" = $1 AND "public"."GeoConfig"."state" = $2) AND 1=1) LIMIT $3 OFFSET $4
```

Client engine log:
```
  prisma:client  Prisma Client call: +1s
  prisma:client  prisma.geoConfig.findUnique({
  where: {
    country_state: {
      country: "US",
      state: "AK"
    }
  }
}) +0ms
  prisma:client  Generated request: +0ms
  prisma:client  {
  "modelName": "GeoConfig",
  "action": "findUnique",
  "query": {
    "arguments": {
      "where": {
        "country_state": {
          "country": "US",
          "state": "AK"
        }
      }
    },
    "selection": {
      "$composites": true,
      "$scalars": true
    }
  }
}
```

**Comment by imwexpex** [comment]
Spotted the issue. This happens if Prisma calls are inside the array Map function, and ONLY if the array has more than 1 element in it

<img width="361" alt="image" src="https://github.com/prisma/prisma/assets/25363536/f5c90ea2-6150-47a7-8b3b-49b4a724cef4">
<img width="398" alt="image" src="https://github.com/prisma/prisma/assets/25363536/04635376-d971-4a57-b40a-14a13889456b">

**Comment by imwexpex** [comment]
Prisma client even generates different sql Query

Query when array.length > 1:
```
prisma:query SELECT "public"."GeoConfig"."uuid", "public"."GeoConfig"."country", "public"."GeoConfig"."state", "public"."GeoConfig"."minAge", "public"."GeoConfig"."enabled", "public"."GeoConfig"."createdAt", "public"."GeoConfig"."updatedAt", "public"."GeoConfig"."updatedBy" FROM "public"."GeoConfig" WHERE (("public"."GeoConfig"."country" = $1 AND "public"."GeoConfig"."state" = $2) OR ("public"."GeoConfig"."country" = $3 AND "public"."GeoConfig"."state" = $4)) OFFSET $5
```

Query when array.length === 1:
```
prisma:query SELECT "public"."GeoConfig"."uuid", "public"."GeoConfig"."country", "public"."GeoConfig"."state", "public"."GeoConfig"."minAge", "public"."GeoConfig"."enabled", "public"."GeoConfig"."createdAt", "public"."GeoConfig"."updatedAt", "public"."GeoConfig"."updatedBy" FROM "public"."GeoConfig" WHERE (("public"."GeoConfig"."country" = $1 AND "public"."GeoConfig"."state" = $2) AND 1=1) LIMIT $3 OFFSET $4
```

**Comment by SevInf** [comment]
Can confirm with the following schema:

```prisma
model model {
  id       String   @id @default(uuid())
  category Category @unique
}

enum Category {
  A
  B
}
```

and the code:

```ts
async function main() {
  const prisma = new PrismaClient()
  await prisma.model.deleteMany()
  await prisma.model.create({
    data: { category: 'A' },
  })

  const results = await Promise.all([
    prisma.model.findUnique({ where: { category: 'A' } }),
    prisma.model.findUnique({ where: { category: 'B' } }),
  ])

  console.log(results)
}
```

Happens on 5.0.0 and 4.16.2 with `jsonProtocol`, does not happen with GraphQL.

**Comment by SevInf** [comment]
Hello everyone. The fix is merged and will be a part of 5.1 release. If you want to test it before that, you can try `5.1.0-dev.15` development snapshot. We don't recommend using it in production, but it should be enough to ensure that fix works for you.

**Comment by imwexpex** [comment]
@SevInf thanks for the quick responses and fix! I can confirm that the bug is fixed in the `5.1.0-dev.15` development snapshot.

**Comment by jkepps** [comment]
@SevInf do we have an estimated timeline for the 5.1.0 release?

**Comment by Jolg42** [comment]
@jkepps Yes! 5.1.0 release is planned for Tuesday, August 1st.

**Comment by gkkirsch** [comment]
Yikes, just running into this in production. I thought I was going crazy. We are getting nulls where we are expecting full records. This happens as soon as an enum is including in a index.

**Comment by imwexpex** [comment]
> Yikes, just running into this in production. I thought I was going crazy. We are getting nulls where we are expecting full records. This happens as soon as an enum is included in an index.
> 
> Channel is an ENUM.
> 
> ```
> 
>   @@unique([userId, event, channel])
> 
> ```
> 
> Thanks for jumping on this quickly.
> 
> 
> 
> For anyone getting this bug in production this could cause some pretty bad non intended consequences. In our example we could have sent notifications to all users that had their notifications turned off.
> 
> 
> 
> 
> 
> @SevInf Any chance we can we get a bug patch before August?

You can downgrade to previous Prisma version and disable JSONprotocol as a workaround

**Comment by SevInf** [comment]
@gkkirsch I too would suggest to hold off 5.x update until Tuesday in your case.

**Comment by maiconsanson** [comment]
This issue was closed but still occurs in `5.8.0`.

Using `findUnique` with a @unique enum field doesn't work.
Only with `findFirst`.

**Comment by Jolg42** [comment]
@maiconsanson since this issue is closed, could you create a new issue with what's needed to get to a reproduction ([see the bug report template](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&projects=&template=bug_report.yml))?

**Comment by maiconsanson** [comment]
> This issue was closed but still occurs in `5.8.0`.
> 
> Using `findUnique` with an `@unique` `enum` field doesn't work. Only with `findFirst`.

@Jolg42 I've made new tests and realized that the error was mine. It works as expected, indeed. Sorry for the false alarm.
